File: diffusion/diffusion.f90
!========================================================================
!    This code was written by Li Chen at Xi'an Jiaotong University.          
!    This code was for pure diffusion in a 2D channel. 
!    Concentrations are known at the left inlet and right outlet,
!    and no flux boundary conditions at the top and bottom walls.
!    One can refer to the following papers for more details: 
!          Li Chen et al., International Journal of Thermal Sciences, 51:132-144,2012
!          Li Chen et al., Physical Review E, 87(4):043306,2013
!    Li Chen: lichennht08@mail.xjtu.edu.cn. 
!========================================================================
MODULE START_L
    PARAMETER (nx=61,ny=21)  
    integer*4::i,j,k,iter
    double precision,parameter::xl=60.E-6,yl=20.E-6  
!--------------------------------------------------------------------
    double precision::dx,dy,dt
    integer*4,dimension(0:4)::fcx=(/0,1,0,-1,0/)
    integer*4,dimension(0:4)::fcy=(/0,0,1,0,-1/)
!--------------------------------------------------------------------
    double precision,dimension(nx,ny)::con,u,v
    double precision,dimension(0:4,0:nx+1,0:ny+1)::g,gg
    double precision::dtao,ts,geq,diff_phy,diff_lat
    double precision::con_in,con_out
!--------------------------------------------------------------------
    logical, dimension(0:nx+1,0:ny+1)::walls
    integer*4,dimension(0:nx+1,0:ny+1)::ls
   double precision::sumc_last,sumc
    double precision::delta
!--------------------------------------------------------------
 END MODULE 
!========================================================================

!========================================================================
PROGRAM MAIN
USE START_L

    dx=xl/float(nx-1)
    dy=dx

    CALL SOLID_STRUCTURE
    CALL INITIALIZATION

    diff_phy=1.e-5 ! unit m^2/s diffusivity in physical unit
    dtao=1.d0       
    ts=0.2d0
    diff_lat=(1.d0-ts)*(dtao-0.5d0)/2.d0 !diffusivity in lattice unit
    scale=diff_phy/diff_lat
    dt=dx**2.d0/scale  ! physical time per lattice step

    last=200000
    delta=1.d0 
       	   	    
    do iter=1,last
        CALL COLLISIOND     
        CALL STREAMD               
        CALL BOUNDARYD             
        CALL MACROD
        if(mod(iter,1000).eq.0) CALL OUTPUT
    enddo	    

END PROGRAM
!========================================================================

!========================================================================
SUBROUTINE INITIALIZATION
USE START_L

    u=0.d0
    v=0.d0
    con=0.d0
    con_in=1.d0
    con_out=0.d0
    do j=1,ny
    do i=1,nx
        do k=0,4
            temp_uv=fcx(k)*u(i,j)+fcy(k)*v(i,j)
            if(k.eq.0) then
                geq=con(i,j)*(ts+0.5d0*temp_uv)
            else
                geq=con(i,j)*((1.d0-ts)/4.d0 + 0.5d0*temp_uv)         
            endif   
            gg(k,i,j)=geq
            g(k,i,j)=geq
        enddo      
    enddo
    enddo

RETURN
END SUBROUTINE
!========================================================================

!========================================================================
SUBROUTINE SOLID_STRUCTURE
USE START_L

    ! ls represents the porous structure: 0 denotes nodes of void space, 1 denotes solid node.
    ls=0   
    ls(:,0:1)=1
    ls(:,ny:ny+1)=1   
     
    walls=.false. 
    do j=0,ny+1
    do i=0,nx+1    
        if(ls(i,j).eq.1) then
            walls(i,j)=.true.
        endif
    enddo
    enddo 
   
RETURN
END SUBROUTINE
!========================================================================

!========================================================================
SUBROUTINE COLLISIOND
USE START_L

    do j=1,ny
    do i=1,nx	
        if(.not.walls(i,j)) then             
            do k=0,4
                temp_uv=fcx(k)*u(i,j)+fcy(k)*v(i,j)
                if(k.eq.0) then
                    geq=con(i,j)*(ts+0.5d0*temp_uv)
                else
                    geq=con(i,j)*((1.-ts)/4.d0 + 0.5d0*temp_uv)         
                endif   
                    gg(k,i,j)=g(k,i,j)-1.d0/dtao*(g(k,i,j)-geq)
                ! if(gg(k,i,j).gt.0.1d0) then
                !     write(*,*) i,j,k,gg(k,i,j)
                ! endif
            enddo      
        endif
    enddo
    enddo
    

RETURN
END SUBROUTINE
!========================================================================

!========================================================================
SUBROUTINE STREAMD
USE START_L
 
	do j=1,ny
	do i=1,nx
	  g(0,i,j)=gg(0,i,j)
	  g(1,i,j)=gg(1,i-1,j)
	  g(3,i,j)=gg(3,i+1,j)
	  g(2,i,j)=gg(2,i,j-1)
	  g(4,i,j)=gg(4,i,j+1)	  
	enddo
    enddo
        
RETURN 
END SUBROUTINE
!========================================================================

!========================================================================
SUBROUTINE BOUNDARYD
USE START_L

    do j=1,ny   
    do i=1,nx		 
         if(walls(i,j)) then		   
		    gg(1,i,j)=g(3,i,j)
		    gg(3,i,j)=g(1,i,j)
		    gg(2,i,j)=g(4,i,j)
		    gg(4,i,j)=g(2,i,j)
         endif		   
    enddo
    enddo

    do i=1,nx
        gg(2,i,1)=g(4,i,1)  ! bounce back for noflux boundary condition
        gg(4,i,ny)=g(2,i,ny) !bounce back for noflux boundary condition
    enddo

    do j=1,ny
        g(1,1,j)=con_in-g(0,1,j)-g(2,1,j)-g(3,1,j)-g(4,1,j) !concentration boundary condition
        g(3,nx,j)=con_out-g(0,nx,j)-g(1,nx,j)-g(2,nx,j)-g(4,nx,j) !concentration boundary condition
    enddo
                
RETURN 
END SUBROUTINE
!========================================================================

!========================================================================
SUBROUTINE MACROD
USE START_L
	
    do j=1,ny
    do i=1,nx
        if(.not.walls(i,j)) then
            con(i,j)=g(0,i,j)+g(1,i,j)+g(2,i,j)+g(3,i,j)+g(4,i,j)
        endif
    enddo
    enddo
	   
    do j=1,ny
        con(1,j)=con_in
        con(nx,j)=con_out
    enddo

RETURN
END SUBROUTINE
!========================================================================

!========================================================================
SUBROUTINE OUTPUT
USE START_L

    sumc_last=sumc
    sumc=0.d0
    do j=1,ny
    do i=1,nx
        if(.not.walls(i,j)) sumc=sumc+con(i,j)
    enddo
    enddo             
    delta=abs(sumc_last-sumc)/abs(sumc) 
    write(*,*) iter,con(nx/2,ny/2),delta,sumc
       
    open(10,file="concentration.dat")
    write(10,*)'VARIABLES= X,Y,con' 
    WRITE(10,*)'ZONE I=',nx,',J=',ny-2,',T=TT'
    do j=2,ny-1
    do i=1,nx
        write(10,*) i,j,con(i,j)
    enddo
    enddo
    close(10)

    open(10,file="concentration_x.dat")
    do i=1,nx
        write(10,*) i,con(i,ny/2)
    enddo
    close(10)

    if(delta.le.1.e-8) stop
    ! stop

RETURN
END SUBROUTINE
!========================================================================

File: diffusion/diffusion.cpp
#include <math.h>

#include <fstream>
#include <iomanip>
#include <iostream>

const int NX = 1020, NY = 712;
int iter;
const double XL = 20.E-6, YL = 20.E-6;
const double DX = XL / float(NX - 1), DY = DX;
const double DIFF_PHY = 1.e-5, DTAO = 1.5, TS = 0.2,
             DIFF_LAT = (1. - TS) * (DTAO - 0.5) / 2.,
             SCALE = DIFF_PHY / DIFF_LAT, dt = pow(DX, 2.) / SCALE;
double u[NY + 1][NX + 1] = {0.}, v[NY + 1][NX + 1] = {0.},
                      con[NY + 2][NX + 2] = {0.};
const int GCX[9] = {0, 1, 0, -1, 0, 1, -1, -1, 1};
const int GCY[9] = {0, 0, 1, 0, -1, 1, 1, -1, -1};
// const double w[9] = {4. / 9.,  1. / 9.,  1. / 9.,  1. / 9., 1. / 9.,
//                      1. / 36., 1. / 36., 1. / 36., 1. / 36.};
const double w[5] = {1. / 2., 1. / 4., 1. / 4., 1. / 4., 1. / 4.};
double g[NY + 2][NX + 2][9], gg[NY + 2][NX + 2][9];
double geq;
const double CON_IN = 1., CON_OUT = 0.;
int ls[NY + 2][NX + 2] = {0};
bool walls[NY + 2][NX + 2] = {false};
const double DELTA = 1.;
int last = 200000;
double sumc_last, sumc;

void imread(std::string filename) {
  std::ifstream imgfile(filename);
  if (!imgfile.is_open()) {
    std::cout << "can not open this file" << std::endl;
    return;
  }
  for (int y = 1; y < NY + 1; y++) {
    for (int x = 1; x < NX + 1; x++) {
      float buf;
      imgfile >> buf;
      ls[y][x] = (int)buf;
    }
  }
}
void solid_structure() {
  // imread("masks/xct_0414_left.txt");
  for (int j = 0; j < NY + 2; j++) {
    ls[j][0] = 0;
    ls[j][1] = 0;
    ls[j][2] = 0;
    ls[j][NX - 1] = 0;
    ls[j][NX] = 0;
    ls[j][NX + 1] = 0;
  }
  for (int i = 0; i < NX + 2; i++) {
    ls[0][i] = 1;
    ls[1][i] = 1;
    ls[NY][i] = 1;
    ls[NY + 1][i] = 1;
  }
  for (int j = 0; j < NY + 2; j++) {
    for (int i = 0; i < NX + 2; i++) {
      if (ls[j][i] == 1) {
        walls[j][i] = true;
      }
    }
  }
}

void initialization() {
  for (int j = 1; j < NY + 1; j++) {
    for (int i = 1; i < NX + 1; i++) {
      for (int k = 0; k < 5; k++) {
        double temp_uv = GCX[k] * u[j][i] + GCY[k] * v[j][i];
        if (k == 0) {
          geq = con[j][i] * (TS + w[0] * temp_uv);
        } else {
          geq = con[j][i] * ((1. - TS) * w[k] + w[0] * temp_uv);
        }
        gg[j][i][k] = geq;
        g[j][i][k] = geq;
      }
    }
  }
}

void collisiond() {
  for (int j = 1; j < NY + 1; j++) {
    for (int i = 1; i < NX + 1; i++) {
      if (!walls[j][i]) {
        for (int k = 0; k < 5; k++) {
          const double temp_uv = GCX[k] * u[j][i] + GCY[k] * v[j][i];
          if (k == 0) {
            geq = con[j][i] * (TS + w[0] * temp_uv);
          } else {
            geq = con[j][i] * ((1. - TS) * w[k] + w[0] * temp_uv);
          }
          gg[j][i][k] = g[j][i][k] - 1. / DTAO * (g[j][i][k] - geq);
          // if (gg[j][i][k] > 0.1) {
          //   std::cout << i << "," << j << "," << k << "," << gg[j][i][k] <<
          //   ","
          //             << std::endl;
          // }
        }
      }
    }
  }
}

void streamd() {
  for (int j = 1; j < NY + 1; j++) {
    for (int i = 1; i < NX + 1; i++) {
      for (int k = 0; k < 5; k++) {
        g[j][i][k] = gg[j - GCY[k]][i - GCX[k]][k];
      }
    }
  }
}

void boundaryd() {
  for (int j = 1; j < NY + 1; j++) {
    for (int i = 1; i < NX + 1; i++) {
      if (walls[j][i]) {
        gg[j][i][1] = g[j][i][3];
        gg[j][i][3] = g[j][i][1];
        gg[j][i][2] = g[j][i][4];
        gg[j][i][4] = g[j][i][2];
        // gg[j][i][5] = g[j][i][7];
        // gg[j][i][7] = g[j][i][5];
        // gg[j][i][6] = g[j][i][8];
        // gg[j][i][8] = g[j][i][6];
      }
    }
  }
  for (int i = 1; i < NX + 1; i++) {
    gg[1][i][2] = g[1][i][4];
    gg[NY][i][4] = g[NY][i][2];
    // gg[1][i][5] = g[1][i][8];
    // gg[NY][i][8] = g[NY][i][5];
    // gg[1][i][6] = g[1][i][7];
    // gg[NY][i][7] = g[NY][i][6];
  }
  for (int j = 1; j < NY + 1; j++) {
    g[j][1][1] = CON_IN;
    g[j][NX][3] = CON_OUT;
    for (int k = 0; k < 5; k++) {
      if (k != 1) {
        g[j][1][1] -= g[j][1][k];
      }
      if (k != 3) {
        g[j][NX][3] -= g[j][NX][k];
      }
    }
  }
}

void macrod() {
  for (int j = 1; j < NY + 1; j++) {
    for (int i = 1; i < NX + 1; i++) {
      if (!walls[j][i]) {
        con[j][i] = 0.;
        for (int k = 0; k < 5; k++) {
          con[j][i] += g[j][i][k];
        }
      }
    }
  }
  for (int j = 1; j < NY + 1; j++) {
    con[j][1] = CON_IN;
    con[j][NX] = CON_OUT;
  }
}

void output() {
  const double sumc_last = sumc;
  sumc = 0.;
  for (int j = 1; j < NY + 1; j++) {
    for (int i = 1; i < NX + 1; i++) {
      // if (!walls[j][i]) {
      sumc += con[j][i];
      // }
    }
  }
  const double delta = abs(sumc_last - sumc) / abs(sumc);
  // std::cout << sumc_last << "," << sumc << "," << delta << std::endl;
  std::cout << iter << "," << iter * dt << "," << con[NY / 2][NX / 2] << ","
            << delta << "," << sumc << std::endl;
  // std::ofstream outfile("data.csv", std::ios::app);
  // outfile << iter * dt << "," << con[NY / 2][NX / 2] << std::endl;
  // outfile.close();
  // double c = 0;
  // for (int y = 2; y < NY / 5; y++) {
  //   for (int x = 2; x < NX - 2; x++) {
  //     c += con[y][x];
  //   }
  // }
  // c /= (NX - 2 - 2) * (NY / 5 - 2);
  // std::cout << c << std::endl;
  if (delta < 1.e-8) {
    std::ofstream outfile("concentration_cpp.dat", std::ios::out);
    outfile << "VARIABLES= X,Y,u,v,pre" << std::endl;
    outfile << "ZONE I=" << NX << ",J=" << NY << ",T=TT" << std::endl;
    for (int j = 2; j < NY; j++) {
      for (int i = 1; i < NX + 1; i++) {
        outfile << i << "," << j << "," << u[j][i] << "," << v[j][i] << ","
                << con[j][i] << std::endl;
      }
    }
    outfile.close();

    last = 0;
  }
  // last = 0;
}

int main() {
  solid_structure();
  initialization();
  for (iter = 1; iter < last + 1; iter++) {
    collisiond();
    streamd();
    boundaryd();
    macrod();
    if (iter % 100 == 0) {
      output();
    }
  }
}